### PR TITLE
bmx7: transfer package from routing

### PR DESCRIPTION
**Pull request by aparcar**
Maintainer: @axn 
Compile tested: ath79/generic
Run tested: -

Description:

Also updating the Makefile to the the new format used in packages.git.
Future releases should use a tag instead of a commit hash.

Signed-off-by: Paul Spooren <mail@aparcar.org>

### CODE REVIEW

File: net/bmx7/Makefile [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+PKG_NAME:=bmx7
+PKG_VERSION:=cc245a22e54b5e110384485f18d9b805c91d8b71
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://codeload.github.com/bmx-routing/bmx7/tar.gz/$(PKG_VERSION)?
+PKG_HASH:=bc68f2ce8ff9c435deff31833bdb39eafa3aeda1c19809705e2b4b1f1d4c8256
+
+PKG_MAINTAINER:=Axel Neumann <neumann@cgws.de>
+PKG_LICENSE:=GPL-2.0
```
Comment (neheb) [inline net/bmx7/Makefile]: This should be -or-later
PKG_LICENSE_FILES should be LICENSE

Comment (aparcar) [reply net/bmx7/Makefile]: Is that actually a requirement?

Comment (neheb) [reply net/bmx7/Makefile]: We want consistency between packages. From a quick look, this is -or-later.

Comment (neheb) [reply net/bmx7/Makefile]: Put another way, the unofficial standard for PKG_LICENSE is SPDX tags. See: https://spdx.org/licenses/

File: net/bmx7/Makefile [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        CC="$(TARGET_CC)" \
+        INSTALL_DIR="$(PKG_INSTALL_DIR)" \
+        build_all
+
+define Package/bmx7/Default
+  SECTION:=net
+  CATEGORY:=Network
+  SUBMENU:=Routing and Redirection
+  TITLE:=BMX7 layer 3 routing daemon
+  URL:=http://github.com/bmx-routing/bmx7
```
Comment (BKPepe) [inline net/bmx7/Makefile]:
```suggestion
  URL:=https://github.com/bmx-routing/bmx7
```

File: net/bmx7/Makefile (hunk 1 above)
Comment (aparcar) [reply net/bmx7/Makefile]: updated, thanks

File: net/bmx7/Makefile (hunk 2 above)
Comment (aparcar) [reply net/bmx7/Makefile]: updated, thanks

**Comment by aparcar** [comment]
Updated bmx7 to 7.1.0 instead of using the commit hash.

File: net/bmx7/Makefile [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+define Package/bmx7-uci-config
+  $(call Package/bmx7/Default)
+  DEPENDS:=bmx7 +libuci
+  TITLE:=configuration plugin based on uci (recommended!)
+endef
+
+define Package/bmx7-iwinfo
+  $(call Package/bmx7/Default)
+  DEPENDS:=bmx7 +libiwinfo
+  TITLE:=link characteristics plugin based on libiwinfo (recommended!)
```
Comment (BKPepe) [inline net/bmx7/Makefile]: This title is so long that it isn't shown in `make menuconfig`.

File: net/bmx7/Makefile [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  SECTION:=net
+  CATEGORY:=Network
+  SUBMENU:=Routing and Redirection
+  TITLE:=BMX7 layer 3 routing daemon
+  URL:=https://github.com/bmx-routing/bmx7
+  DEPENDS:=+zlib +libmbedtls +libiwinfo
+endef
+
+define Package/bmx7/description
+BMX7 routing daemon supporting securely-entrusted IPv6 (and IPv4in6) routing
```
Comment (BKPepe) [inline net/bmx7/Makefile]:
```suggestion
  BMX7 routing daemon supporting securely-entrusted IPv6 (and IPv4in6) routing.
```

**Comment by aparcar** [comment]
Updated, thanks!

File: net/bmx7/files/etc/init.d/bmx7 [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh /etc/rc.common
+START=91
+USE_PROCD=1
+
+BIN=/usr/sbin/bmx7
+CONF=/etc/config/bmx7
+
+start_service() {
+	cd /root/
```
Comment (neheb) [inline net/bmx7/files/etc/init.d/bmx7]: Shellcheck reports this error: https://github.com/koalaman/shellcheck/wiki/SC2164

Just add || exit

Comment (aparcar) [reply net/bmx7/files/etc/init.d/bmx7]: Thanks, updated

File: net/bmx7/files/etc/init.d/bmx7 [hunk 6]
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh /etc/rc.common
+START=91
```
Comment (diizzyy) [inline net/bmx7/files/etc/init.d/bmx7]: Any reason why this is 91 instead of 99?

File: net/bmx7/Makefile [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+define Package/bmx7
+  $(call Package/bmx7/Default)
+  MENU:=1
+endef
+
+define Package/bmx7-uci-config
+  $(call Package/bmx7/Default)
+  DEPENDS:=bmx7 +libuci
+  TITLE:=configuration plugin based on uci (recommended!)
```
Comment (diizzyy) [inline net/bmx7/Makefile]: Remove ```(recommended!)``` and capitalize titles

File: net/bmx7/Makefile [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+endef
+
+define Package/bmx7-table
+  $(call Package/bmx7/Default)
+  DEPENDS:=bmx7 +bmx7-tun
+  TITLE:=plugin to announce routes from tables via tunnels
+endef
+
+define Build/Configure
+	mkdir -p $(PKG_INSTALL_DIR)
```
Comment (diizzyy) [inline net/bmx7/Makefile]: Why is this needed?

File: net/bmx7/Makefile [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+PKG_LICENSE:=GPL-2.0-or-later
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_BUILD_PARALLEL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+TARGET_CFLAGS += $(FPIC)
+
+MAKE_ARGS += EXTRA_CFLAGS="$(TARGET_CFLAGS) -I. -I$(STAGING_DIR)/usr/include -DCRYPTLIB=MBEDTLS_2_8_0 -DCORE_LIMIT=20000 -DTRAFFIC_DUMP -DNO_TRACE_FUNCTION_CALLS -DBMX7_LIB_IWINFO"
```
Comment (diizzyy) [inline net/bmx7/Makefile]: Why is this not in the same variable?

**Comment by diizzyy** [comment]
Several FLAGS are (incorrectly) passed multiple times and hardset, this needs to be fixed (preferably upstream if possible).

Sample:
```mips-openwrt-linux-musl-gcc  -pedantic -W -Wall -Wstrict-prototypes -Wno-unused-parameter -Os -g3 -std=gnu99 -DGIT_REV=\"\" -DAVL_5XLINKED -DDEBUG_MALLOC -DCORE_LIMIT=20000      -pedantic -W -Wall -Wstrict-prototypes -Wno-unused-parameter -Os -g3 -std=gnu99 -DGIT_REV=\"\" -DAVL_5XLINKED -DDEBUG_MALLOC -DCORE_LIMIT=20000      -Os -pipe -mno-branch-likely -mips32r2 -mtune=24kc -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -msoft-float -mips16 -minterlink-mips16 -iremap/home/build/build_dir/build_dir/target-mips_24kc_musl/bmx7-7.1.0:bmx7-7.1.0 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -fpic -I. -I/home/build/build_dir/staging_dir/target-mips_24kc_musl/usr/include -DCRYPTLIB=MBEDTLS_2_8_0 -DCORE_LIMIT=20000 -DTRAFFIC_DUMP -DNO_TRACE_FUNCTION_CALLS -DBMX7_LIB_IWINFO -c node.c -o node.o```

Examples:
https://github.com/bmx-routing/bmx7/blob/master/Common.mk#L3
https://github.com/bmx-routing/bmx7/blob/master/Common.mk#L34
https://github.com/bmx-routing/bmx7/blob/master/Common.mk#L65
https://github.com/bmx-routing/bmx7/blob/master/lib/bmx7_evil/Makefile#L2

Since encryption is enforced you might want to disable MIPS16 instructions

**Comment by diizzyy** [comment]
@aparcar 
Ping

**Comment by aparcar** [comment]
@axn please look at the TARGET_CFLAGS and MAKE_ARGS issue, I can't tell what you are doing there.

File: net/bmx7/Makefile [hunk 10]
```diff
@@ -0,0 +1,165 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=bmx7
+PKG_VERSION:=7.1.1
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://codeload.github.com/bmx-routing/bmx7/tar.gz/v$(PKG_VERSION)?
+PKG_HASH:=5f88df1c95e5cb842a6016bb1604e3e7f6097c63c5c9916edc3c84e96d4f5f65
+PKG_UNPACK:=$(TAR) -C $(BUILD_DIR)/$(PKG_NAME)-$(PKG_VERSION) \
```
Comment (neheb) [inline net/bmx7/Makefile]: Can be replaced with PKG_BUILD_DIR

Comment (aparcar) [reply net/bmx7/Makefile]: I tried but apparently a folder named `src/` is treated differently, I solved it via a prepare statement.

Comment (neheb) [reply net/bmx7/Makefile]: Ah yes you're right. There was a subdirectory make flag. For CMake, this is CMAKE_SOURCE_SUBDIR. I'll look into this later

Comment (aparcar) [reply net/bmx7/Makefile]: @neheb Hey you came up with something?

Comment (neheb) [reply net/bmx7/Makefile]: MAKE_PATH:=src is what you want I think.

Comment (aparcar) [reply net/bmx7/Makefile]: Works, thank you!

**Comment by neheb** [comment]
Please check @diizzyy comments

**Comment by aparcar** [comment]
I can't solve the flag issue as I'm not familiar enough in C nor bmx7, closing this for now.

**Comment by BKPepe** [comment]
There are missing references to the routing repository, where the package was updated, and the Makefile was refactored a little bit by the OP.

- https://github.com/openwrt/routing/pull/507
- https://github.com/openwrt/routing/pull/512
